File: scio-schemas/src/main/scala/com/spotify/scio/avro/AvroUtils.scala
/*
 * Copyright 2016 Spotify AB.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package com.spotify.scio.avro

import org.apache.avro.Schema
import org.apache.avro.generic.{GenericData, GenericRecord}

import scala.annotation.switch
import scala.collection.JavaConverters._

object AvroUtils {

  private def f(name: String, tpe: Schema.Type) =
    new Schema.Field(
      name,
      Schema.createUnion(List(Schema.create(Schema.Type.NULL), Schema.create(tpe)).asJava),
      null: String,
      null: AnyRef)

  private def fArr(name: String, tpe: Schema.Type) =
    new Schema.Field(name, Schema.createArray(Schema.create(tpe)), null: String, null: AnyRef)

  val schema = Schema.createRecord("GenericTestRecord", null, null, false)
  schema.setFields(
    List(
      f("int_field", Schema.Type.INT),
      f("long_field", Schema.Type.LONG),
      f("float_field", Schema.Type.FLOAT),
      f("double_field", Schema.Type.DOUBLE),
      f("boolean_field", Schema.Type.BOOLEAN),
      f("string_field", Schema.Type.STRING),
      fArr("array_field", Schema.Type.STRING)
    ).asJava)

  def newGenericRecord(i: Int): GenericRecord = {
    val r = new GenericData.Record(schema)
    r.put("int_field", 1 * i)
    r.put("long_field", 1L * i)
    r.put("float_field", 1F * i)
    r.put("double_field", 1.0 * i)
    r.put("boolean_field", true)
    r.put("string_field", "hello")
    r.put("array_field", List[CharSequence]("a", "b", "c").asJava)
    r
  }

  def newSpecificRecord(i: Int): TestRecord =
    new TestRecord(i,
                   i.toLong,
                   i.toFloat,
                   i.toDouble,
                   true,
                   "hello",
                   List[CharSequence]("a", "b", "c").asJava)

  def newCaseClassSpecificRecord(i: Int): CaseClassTestRecord =
    new CaseClassTestRecord(Option(i),
                            Option(i.toLong),
                            Option(i.toFloat),
                            Option(i.toDouble),
                            Option(true),
                            Option("hello"),
                            List[String]("a", "b", "c"))

}

// scalastyle:off
case class CaseClassTestRecord(var int_field: Option[Int] = None,
                               var long_field: Option[Long] = None,
                               var float_field: Option[Float] = None,
                               var double_field: Option[Double] = None,
                               var boolean_field: Option[Boolean] = None,
                               var string_field: Option[String] = None,
                               var array_field: Seq[String] = Seq())
    extends org.apache.avro.specific.SpecificRecordBase {
  def this() = this(None, None, None, None, None, None, Seq())
  def get(field$ : Int): AnyRef = {
    (field$ : @switch) match {
      case 0 => {
        int_field match {
          case Some(x) => x
          case None    => null
        }
      }.asInstanceOf[AnyRef]
      case 1 => {
        long_field match {
          case Some(x) => x
          case None    => null
        }
      }.asInstanceOf[AnyRef]
      case 2 => {
        float_field match {
          case Some(x) => x
          case None    => null
        }
      }.asInstanceOf[AnyRef]
      case 3 => {
        double_field match {
          case Some(x) => x
          case None    => null
        }
      }.asInstanceOf[AnyRef]
      case 4 => {
        boolean_field match {
          case Some(x) => x
          case None    => null
        }
      }.asInstanceOf[AnyRef]
      case 5 => {
        string_field match {
          case Some(x) => x
          case None    => null
        }
      }.asInstanceOf[AnyRef]
      case 6 => {
        scala.collection.JavaConverters
          .bufferAsJavaListConverter({
            array_field map { x =>
              x
            }
          }.toBuffer)
          .asJava
      }.asInstanceOf[AnyRef]
      case _ => new org.apache.avro.AvroRuntimeException("Bad index")
    }
  }
  def put(field$ : Int, value: Any): Unit = {
    (field$ : @switch) match {
      case 0 =>
        this.int_field = {
          value match {
            case null => None
            case _    => Some(value)
          }
        }.asInstanceOf[Option[Int]]
      case 1 =>
        this.long_field = {
          value match {
            case null => None
            case _    => Some(value)
          }
        }.asInstanceOf[Option[Long]]
      case 2 =>
        this.float_field = {
          value match {
            case null => None
            case _    => Some(value)
          }
        }.asInstanceOf[Option[Float]]
      case 3 =>
        this.double_field = {
          value match {
            case null => None
            case _    => Some(value)
          }
        }.asInstanceOf[Option[Double]]
      case 4 =>
        this.boolean_field = {
          value match {
            case null => None
            case _    => Some(value)
          }
        }.asInstanceOf[Option[Boolean]]
      case 5 =>
        this.string_field = {
          value match {
            case null => None
            case _    => Some(value.toString)
          }
        }.asInstanceOf[Option[String]]
      case 6 =>
        this.array_field = {
          value match {
            case (array: java.util.List[_]) => {
              Seq(
                scala.collection.JavaConverters
                  .asScalaIteratorConverter(array.iterator)
                  .asScala
                  .toSeq map { x =>
                  x.toString
                }: _*)
            }
          }
        }.asInstanceOf[Seq[String]]
      case _ => new org.apache.avro.AvroRuntimeException("Bad index")
    }
    ()
  }
  def getSchema: org.apache.avro.Schema = CaseClassTestRecord.SCHEMA$
}

object CaseClassTestRecord {
  val SCHEMA$ = new org.apache.avro.Schema.Parser().parse("{\"type\":\"record\"," +
    "\"name\":\"CaseClassTestRecord\",\"namespace\":\"com.spotify.scio.avro\",\"doc\":\"Record" +
    " for testing\",\"fields\":[{\"name\":\"int_field\",\"type\":[\"null\",\"int\"],\"default\":null},{\"name\":\"long_field\",\"type\":[\"null\",\"long\"],\"default\":null},{\"name\":\"float_field\",\"type\":[\"null\",\"float\"],\"default\":null},{\"name\":\"double_field\",\"type\":[\"null\",\"double\"],\"default\":null},{\"name\":\"boolean_field\",\"type\":[\"null\",\"boolean\"],\"default\":null},{\"name\":\"string_field\",\"type\":[\"null\",\"string\"],\"default\":null},{\"name\":\"array_field\",\"type\":{\"type\":\"array\",\"items\":\"string\"},\"default\":null}]}")
}

// scalastyle:on

File: scio-smb/src/main/java/org/apache/beam/sdk/extensions/smb/AvroUtils.java
/*
 * Copyright 2021 Spotify AB.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package org.apache.beam.sdk.extensions.smb;

import com.google.common.base.Preconditions;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.nio.ByteBuffer;
import java.util.Collection;
import java.util.Map;
import org.apache.avro.Schema;
import org.apache.avro.generic.GenericData;
import org.apache.avro.generic.GenericRecord;
import org.apache.avro.reflect.ReflectData;
import org.apache.beam.sdk.coders.AtomicCoder;
import org.apache.beam.sdk.coders.ByteArrayCoder;
import org.apache.beam.sdk.coders.Coder;
import org.apache.beam.sdk.coders.CoderException;
import org.apache.beam.sdk.coders.StringUtf8Coder;
import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.collect.ImmutableMap;

class AvroUtils {
  private AvroUtils() {}

  /**
   * Constructs the sequence of indexes to access the nested key field from an avro {@link
   * org.apache.avro.generic.IndexedRecord}.
   *
   * @param keyField name of the field (joined with '.')
   * @param keyClass key class to ensure type correctness of the designated keyField
   * @param schema avro schema
   * @return sequence of index to access the keyField
   */
  public static int[] toKeyPath(String keyField, Class<?> keyClass, Schema schema) {
    final String[] fields = keyField.split("\\.");
    final int[] path = new int[fields.length];

    Schema cursor = schema;
    for (int i = 0; i < fields.length - 1; i++) {
      final Schema.Field field = cursor.getField(fields[i]);
      Preconditions.checkNotNull(
          field, String.format("Key path %s does not exist in schema %s", fields[i], cursor));

      cursor = getSchemaOrInnerUnionSchema(field.schema());
      Preconditions.checkArgument(
          cursor.getType() == Schema.Type.RECORD,
          "Non-leaf key field " + fields[i] + " is not a Record type");
      path[i] = field.pos();
    }

    final Schema.Field finalKeyField = cursor.getField(fields[fields.length - 1]);
    Preconditions.checkNotNull(
        finalKeyField,
        String.format(
            "Leaf key field %s does not exist in schema %s", fields[fields.length - 1], cursor));

    final Class<?> finalKeyFieldClass = getKeyClassFromSchema(finalKeyField.schema());

    Preconditions.checkArgument(
        finalKeyFieldClass.isAssignableFrom(keyClass),
        String.format(
            "Key class %s did not conform to its Avro schema. Must be of class: %s",
            keyClass, finalKeyFieldClass));
    path[fields.length - 1] = finalKeyField.pos();

    return path;
  }

  public static String validateKeyField(String keyField, Class<?> keyClass, Schema schema) {
    toKeyPath(keyField, keyClass, schema);
    return keyField;
  }

  public static String validateKeyField(String keyField, Class<?> keyClass, Class<?> recordClass) {
    final Schema schema = new ReflectData(recordClass.getClassLoader()).getSchema(recordClass);
    toKeyPath(keyField, keyClass, schema);
    return keyField;
  }

  private static Class<?> getKeyClassFromSchema(Schema schema) {
    Schema.Type schemaType = getSchemaOrInnerUnionSchema(schema).getType();

    return getClassForType(schemaType);
  }

  // return the passed schema unless it is a union type, in this case, it will
  // return the non null type and will fail if there are multiple non-null types.
  private static Schema getSchemaOrInnerUnionSchema(Schema schema) {
    Schema returnSchema = schema;

    if (schema.getType() == Schema.Type.UNION) {
      boolean hasNull = false;

      for (Schema typeSchema : schema.getTypes()) {
        if (typeSchema.getType() == Schema.Type.NULL) {
          hasNull = true;
        } else {
          returnSchema = typeSchema;
        }
      }

      Preconditions.checkArgument(
          hasNull && schema.getTypes().size() == 2,
          "A union can only be used as a key field if it contains exactly 1 null "
              + "type and 1 key-class type. Was: "
              + schema);
    }

    return returnSchema;
  }

  private static Class<?> getClassForType(Schema.Type schemaType) {
    switch (schemaType) {
      case RECORD:
        return GenericRecord.class;
      case NULL:
        throw new IllegalArgumentException("Key field cannot be of type Null");
      case ARRAY:
        return Collection.class;
      case BOOLEAN:
        return Boolean.class;
      case MAP:
        return Map.class;
      case ENUM:
        return String.class;
      case INT:
        return Integer.class;
      case DOUBLE:
        return Double.class;
      case LONG:
        return Long.class;
      case STRING:
        return CharSequence.class;
      case FIXED:
        return GenericData.Fixed.class;
      case FLOAT:
        return Float.class;
      case BYTES:
        return ByteBuffer.class;
      default:
        throw new IllegalStateException("Can't match key field schema type " + schemaType);
    }
  }

  // Coders for types commonly used as keys in Avro

  public static Map<Class<?>, Coder<?>> coderOverrides() {
    return ImmutableMap.of(
        ByteBuffer.class, ByteBufferCoder.of(),
        CharSequence.class, CharSequenceCoder.of());
  }

  private static class ByteBufferCoder extends AtomicCoder<ByteBuffer> {
    private static final ByteBufferCoder INSTANCE = new ByteBufferCoder();

    private ByteBufferCoder() {}

    public static ByteBufferCoder of() {
      return INSTANCE;
    }

    @Override
    public void encode(ByteBuffer value, OutputStream outStream)
        throws CoderException, IOException {
      byte[] bytes = new byte[value.remaining()];
      value.get(bytes);
      value.position(value.position() - bytes.length);

      ByteArrayCoder.of().encode(bytes, outStream);
    }

    @Override
    public ByteBuffer decode(InputStream inStream) throws CoderException, IOException {
      return ByteBuffer.wrap(ByteArrayCoder.of().decode(inStream));
    }
  }

  private static class CharSequenceCoder extends AtomicCoder<CharSequence> {
    private static final CharSequenceCoder INSTANCE = new CharSequenceCoder();

    private CharSequenceCoder() {}

    public static CharSequenceCoder of() {
      return INSTANCE;
    }

    @Override
    public void encode(CharSequence value, OutputStream outStream)
        throws CoderException, IOException {
      StringUtf8Coder.of().encode(value.toString(), outStream);
    }

    @Override
    public CharSequence decode(InputStream inStream) throws CoderException, IOException {
      return StringUtf8Coder.of().decode(inStream);
    }
  }
}
